keps/sig-api-machinery/3962-mutating-admission-policies/README.md

```diff
@@ -0,0 +1,1363 @@
+See sigs.k8s.io/structured-merge-diff for more details.
+
+Note: See the alternative consideration section for the alternatives.
+
+Pros:
+- Consistent with Server Side Apply so that we will continue investing SSA as the best way to do patch updates to resources;
+- Does not require the users to learn a new syntax;
+- Inherit the declarative nature;
+- Existing merging strategy to be leverage without rewritten and the effort of maintaining multiple systems;
+- Support the existing makers/openapi extensions.
+
+Cons:
+- Lack of deletion support (see the unsetting values section for the details and workaround);
+- All operations of SSA require the schemas of the incoming objects, which require the schemas to be available from the discovery endpoint or compiled-in. 
+  - For API extensions types like CRDs, this can cause a delay before the policy can start enforcing requests.
+  
+
+### Phase 1
+
+#### API Shape
+Similar to the `validations` field in ValidatingAdmissionPolicy, a `mutations` field will be defined inside MutatingAdmissionPolicy which allows users to define a list of mutations that apply to the specific resources.
+Each mutation field contains a CEL expression which evaluates to a partially populated object representing a Server Side Apply "apply configuration". The apply configuration is then merged into the request object.
+
+Here is an example of injecting an initContainer.
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: MutatingAdmissionPolicy
+metadata:
+  name: "sidecar-policy.example.com"
+spec:
+  paramKind:
+    group: mutations.example.com
+    kind: Sidecar
+    version: v1
+  matchConstraints:
+    resourceRules:
+    - apiGroups:   ["apps"]
+      apiVersions: ["v1"]
+      operations:  ["CREATE"]
+      resources:   ["pods"]
+  matchConditions:
+    - name: does-not-already-have-sidecar
+      expression: "!object.spec.initContainers.exists(ic, ic.name == params.name)"
+  failurePolicy: Fail
+  reinvocationPolicy: IfNeeded
+  mutations:
+    - mutation: >
+        Object{
+          spec: Object.spec{
+            initContainers: [
+              Object.spec.initContainers{
+                name: params.name,
+                image: params.image,
+                args: params.args,
+                restartPolicy: params.restartPolicy
+                // ... other container fields injected here ...
+              }
+            ]
+          }
+        }
+```
+
+For each mutation expression, the expression evaluates to an object that has the same type as the incoming object, and the type alias Object refers to the type (see Type Handling for details).
+
+By using Server Side Apply merge algorithms, schema declarations like
+`x-kubernetes-list-type: map`, that control how a merge is performed, will be
+respected.
+
+However, unlike with server side apply, these mutations will not have a field
+manager. This has important implications in how the merge is performed that will
```

Comment:
We are proposing to do a structural merge here rather than a complete SSA. And there will be no need to involve an field manager. I can further explain offline :)

Reply:
Thanks! Yes I agree that the behavior of ownership should be clear. And I am thinking to follow the way how mutation webhook handles field manager. The basic standard would be not introducing any new/surprising behavior by mutation policy. I will add more details.